OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs

```diff
@@ -231,38 +234,55 @@ public void CloseChat()
 			chatOverlay.Visible = true;
 		}
 
-		public void AddChatLineWrapper(string name, Color nameColor, string text, Color textColor)
+		public void AddChatLineWrapper(ChatLine chatLine)
 		{
-			chatOverlayDisplay?.AddLine(name, nameColor, text, textColor);
+			var chatLineToDisplay = chatLine;
+
+			if (chatLine.IsRepeatable() && chatLine.Equals(lastLine))
+			{
+				repetitions++;
+				chatLineToDisplay = new ChatLine(
+					chatLine.Name,
+					chatLine.NameColor,
+					string.Format("{0} ({1})", chatLine.Text, repetitions + 1),
+					chatLine.TextColor,
+					chatLine.Pool);
+			}
+			else
+				repetitions = 0;
+
+			lastLine = chatLine;
+
+			chatOverlayDisplay?.AddLine(chatLineToDisplay, repetitions > 0);
 
 			// HACK: Force disable the chat notification sound for the in-menu chat dialog
 			// This works around our inability to disable the sounds for the in-game dialog when it is hidden
-			AddChatLine(name, nameColor, text, textColor, chatOverlay == null);
+			AddChatLine(chatLineToDisplay, repetitions > 0, chatOverlay == null);
 		}
 
-		void AddChatLine(string @from, Color nameColor, string text, Color textColor, bool suppressSound)
+		void AddChatLine(ChatLine chatLine, bool isRepeated, bool suppressSound)
```

Comment:
`isRepeated` doesn't sit right with me here either. In practice what this is actually doing is removing the last widget in the scroll panel before adding the new, so wouldn't it be better to just do this explicitly in `AddChatLineWrapper` before calling this?

Reply:
Yeah, `AddChatLineWrapper` is a better place to handle this :+1:

Reply:
I moved the checks for repeated lines to `AddChatLineWrapper`.